android/titanium/src/java/org/appcelerator/titanium/util/TiDownloadManager.java

```diff
@@ -50,19 +60,200 @@ public static TiDownloadManager getInstance()
+			class DownloadOperation implements Runnable
+			{
+				private boolean isDone;
+				private URI uri;
+
+				public DownloadOperation(URI uri)
+				{
+					this.uri = uri;
+				}
+
+				@Override
+				public void run()
+				{
+					try (InputStream stream = blockingDownload(this.uri)) {
+						if (stream != null) {
+							KrollStreamHelper.pump(stream, null);
+						}
+					} catch (Exception ex) {
+					} finally {
+						this.isDone = true;
+					}
+				}
+
+				public boolean isDone()
+				{
+					return this.isDone;
+				}
+			};
+
+			// Perform the blocking download.
+			try {
+				DownloadOperation downloadOperation = new DownloadOperation(uri);
+				(new Thread(downloadOperation)).start();
+				while (downloadOperation.isDone() == false) {
+					Thread.sleep(10);
+				}
+			} catch (Exception ex) {
+			}
+
+			// Return a stream to the downloaded file/content via our response cache.
+			URI cachedUri = TiResponseCache.fetchEndpointFollowingRedirects(uri);
+			if (cachedUri != null) {
+				try {
+					inputStream = TiResponseCache.openCachedStream(cachedUri);
+				} catch (Exception ex) {
+				}
+			}
+			return inputStream;
+		}
+
+		// Convert the given URI to a URL object.
+		// Note: This object will validate the string and throw an exception if malformed.
+		URL url = null;
+		try {
+			url = new URL(uri.toString());
+		} catch (Exception ex) {
+			ex.printStackTrace();
+			return null;
+		}
+
+		// Attempt to download the file.
+		URLConnection connection = null;
+		try {
+			int redirectCount = 0;
+			while (true) {
+				// Attempt to connect to the given URL.
+				connection = url.openConnection();
+				connection.setConnectTimeout(TIMEOUT_IN_MILLISECONDS);
+				connection.setReadTimeout(TIMEOUT_IN_MILLISECONDS);
+				connection.setDoInput(true);
+				if (connection instanceof HttpURLConnection) {
+					// Connect via HTTP/HTTPS.
+					HttpURLConnection httpConnection = (HttpURLConnection) connection;
+					httpConnection.setInstanceFollowRedirects(true);
+					httpConnection.connect();
+					int responseCode = httpConnection.getResponseCode();
+					if (responseCode == 200) {
+						// A valid respoonse has been received. Fetch its download stream.
+						inputStream = httpConnection.getInputStream();
+					} else if ((responseCode >= 300) && (responseCode < 400)) {
+						// A redirect response has been received.
+						// First, make sure we haven't been redirected too many times.
+						// Note: This is a rudimentary check to avoid recursive redirects.
+						redirectCount++;
+						if (redirectCount > 10) {
+							break;
+						}
+
+						// Acquire the URL we're being asked to redirect to.
+						String redirectUrl = httpConnection.getHeaderField("Location");
+						if ((redirectUrl == null) || (redirectUrl.length() <= 0)) {
+							break;
+						}
+						url = new URL(redirectUrl);
+
+						// Close the connection and attempt to connect to the redirect URL.
+						httpConnection.disconnect();
+						continue;
+					}
```

Comment:
Maybe log a warning if there's a bad response (404?), or throw an exception?

Reply:
Good point. The `image` property used to log a 404 warning and now it doesn't. I'll add some log messages.

Also, `backgroundImage` used to crash the app for a 404 response when "run-on-main-thread" was `false`. I've inadvertently fixed that. :)